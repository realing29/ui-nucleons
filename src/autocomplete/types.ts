import type { ReactNode } from 'react';
import type { DropdownProps } from '../dropdown';
import type { DropdownItemElement } from '../dropdown-item/types';
import type { FieldBlockProps } from '../field-block';
import type { InputProps } from '../input';

export interface AutocompleteProps extends InputProps {
  children?: ReactNode;
  loading?: boolean;
  filterOption?: (option: DropdownItemElement, inputValue: string) => boolean;
  adornmentEnd?: FieldBlockProps['adornmentEnd'] | ((data: { menuShown: boolean }) => ReactNode);
  dropdownProps?: Pick<DropdownProps, 'className' | 'style'>;
}
