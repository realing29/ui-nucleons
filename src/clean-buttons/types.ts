/**
 * Размер прозрачных кнопок (высота) по дизайн-гайдам.
 */
export type CleanButtonSize = 's' | 'm' | 'l';
