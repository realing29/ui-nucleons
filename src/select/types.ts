import type {
  ComponentType,
  FocusEventHandler,
  KeyboardEventHandler,
  MouseEventHandler,
  ReactNode,
  RefObject,
} from 'react';
import type { FieldBlockProps, FieldBlockStyle } from '../field-block';
import type { TextButtonProps } from '../text-button';

export interface SelectStyle extends FieldBlockStyle {
  '--dropdown-width'?: string | number;
}

export interface SelectOpenerProps {
  openerRef: RefObject<HTMLElement>;
  label?: string;
  value: ReactNode;
  opened: boolean;
  focused: boolean;
  failed?: boolean;
  disabled?: boolean;
  onFocus: FocusEventHandler<HTMLElement>;
  onBlur: FocusEventHandler<HTMLElement>;
  onMouseDown: MouseEventHandler<HTMLElement>;
  onKeyDown: KeyboardEventHandler<HTMLElement>;
}

export type SelectMenuAlign = 'left' | 'right';

export interface SelectProps {
  /** Опции. */
  children?: ReactNode;

  /** CSS-класс корневого элемента. */
  className?: string;

  /** Отключенное состояние. */
  disabled?: boolean;

  /** Ошибочное состояние. */
  failed?: boolean;

  /** Состояние загрузки опций. */
  loading?: boolean;

  /** Сработает при выборе опции. */
  onValueChange?: (value: string) => void;

  /** Сработает при открытии/закрытии меню. */
  onMenuToggle?: (state: { opened: boolean }) => void;

  /** Стили корневого элемента. */
  style?: SelectStyle;

  /** Значение, выводимое в качестве выбранного. При отсутствии будет выведен label. */
  value?: ReactNode;

  /** Направления выпадания меню опций. */
  menuAlign?: SelectMenuAlign;

  /** Ярлык. */
  label?: string;

  /** Открывающий элемент. */
  opener?: 'field-block' | 'text-button' | ComponentType<SelectOpenerProps>;

  /** Опции FieldBlock. */
  fieldBlockProps?: FieldBlockProps;

  /** Опции TextButton. */
  textButtonProps?: TextButtonProps;
}
