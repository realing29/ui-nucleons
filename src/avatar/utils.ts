import { isString } from 'lodash';

/**
 * Формирует монограмму из переданной строки.
 * @param value Переданная строка.
 * @return Монограмма.
 */
export const getMonogram = (value: any) =>
  isString(value)
    ? value
        .split(/\s+/g)
        .slice(0, 2)
        .map(s => s[0])
        .reverse()
        .join('')
        .toUpperCase()
    : '';
