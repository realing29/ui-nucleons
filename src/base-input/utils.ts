/**
 * Получив объект, вернет его копию но без свойства multiline.
 * @param source Объект.
 * @return Копия без свойства multiline.
 */
export function omitMultiline<T extends Record<string, any>>(source: T): Omit<T, 'multiline'> {
  const next = { ...source };

  delete next.multiline;

  return next;
}
