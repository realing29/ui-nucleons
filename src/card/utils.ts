import { createContext } from 'react';
import { PlateProps } from '../plate';

export const CardContext = createContext<Partial<PlateProps>>({});
