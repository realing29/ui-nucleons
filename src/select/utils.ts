import { createContext, createRef } from 'react';
import { SelectOpenerBinding } from './types';

export const SelectContext = createContext<SelectOpenerBinding>({
  value: '',
  opened: false,
  menuFocused: false,
  openerRef: createRef(),
  anchorRef: createRef(),
  onMouseDown: () => void 0,
  onKeyDown: () => void 0,
});
