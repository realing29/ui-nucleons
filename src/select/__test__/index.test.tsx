import React, { useContext } from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Select } from '..';
import { DropdownItem } from '../../dropdown-item';
import { SelectOpenerProps } from '../types';
import { SelectContext } from '../utils';

describe('Select', () => {
  it('should handle mouse control on opener - FieldBlock', () => {
    const { queryAllByTestId, getByTestId } = render(
      <Select>
        <DropdownItem value='1'>One</DropdownItem>
        <DropdownItem value='2'>Two</DropdownItem>
        <DropdownItem value='3'>Three</DropdownItem>
      </Select>,
    );

    // initial state
    expect(queryAllByTestId('select')).toHaveLength(1);
    expect(queryAllByTestId('field')).toHaveLength(1);
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);

    // opener click
    fireEvent.mouseDown(getByTestId('field:block'));
    expect(queryAllByTestId('select')).toHaveLength(1);
    expect(queryAllByTestId('field')).toHaveLength(1);
    expect(queryAllByTestId('dropdown')).toHaveLength(1);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(3);

    // opener click again
    fireEvent.mouseDown(getByTestId('field:block'));
    expect(queryAllByTestId('select')).toHaveLength(1);
    expect(queryAllByTestId('field')).toHaveLength(1);
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);
  });

  it('should handle mouse control on opener - TextButton', () => {
    const { queryAllByTestId, getByTestId } = render(
      <Select opener='text-button'>
        <DropdownItem value='1'>Foo</DropdownItem>
        <DropdownItem value='2'>Bar</DropdownItem>
      </Select>,
    );

    // initial state
    expect(queryAllByTestId('select')).toHaveLength(1);
    expect(queryAllByTestId('text-button')).toHaveLength(1);
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);

    // opener click
    fireEvent.mouseDown(getByTestId('text-button'));
    expect(queryAllByTestId('select')).toHaveLength(1);
    expect(queryAllByTestId('text-button')).toHaveLength(1);
    expect(queryAllByTestId('dropdown')).toHaveLength(1);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(2);

    // opener click again
    fireEvent.mouseDown(getByTestId('text-button'));
    expect(queryAllByTestId('select')).toHaveLength(1);
    expect(queryAllByTestId('text-button')).toHaveLength(1);
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);
  });

  it('should handle keyboard control on opener - FieldBlock', () => {
    const { queryAllByTestId, getByTestId } = render(
      <Select>
        <DropdownItem value='1'>One</DropdownItem>
        <DropdownItem value='2'>Two</DropdownItem>
        <DropdownItem value='3'>Three</DropdownItem>
      </Select>,
    );

    // initial state
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);

    // opener focus
    fireEvent.focus(getByTestId('field:block'));
    expect(queryAllByTestId('dropdown')).toHaveLength(0);

    // opener enter keydown
    fireEvent.keyDown(getByTestId('field:block'), { key: 'Enter' });
    expect(queryAllByTestId('dropdown')).toHaveLength(1);

    // menu blur
    fireEvent.blur(getByTestId('dropdown'));
    expect(queryAllByTestId('dropdown')).toHaveLength(0);

    // opener focus + enter keydown
    fireEvent.focus(getByTestId('field:block'));
    fireEvent.keyDown(getByTestId('field:block'), { key: 'Enter' });
    expect(queryAllByTestId('dropdown')).toHaveLength(1);

    // opener enter keydown again
    fireEvent.keyDown(getByTestId('field:block'), { key: 'Enter' });
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
  });

  it('should handle keyboard control on opener - TextButton', () => {
    const { queryAllByTestId, getByTestId } = render(
      <Select opener='text-button'>
        <DropdownItem value='1'>One</DropdownItem>
        <DropdownItem value='2'>Two</DropdownItem>
        <DropdownItem value='3'>Three</DropdownItem>
      </Select>,
    );

    // initial state
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);

    // opener focus
    fireEvent.focus(getByTestId('text-button'));
    expect(queryAllByTestId('dropdown')).toHaveLength(0);

    // opener enter keydown
    fireEvent.keyDown(getByTestId('text-button'), { key: 'Enter' });
    expect(queryAllByTestId('dropdown')).toHaveLength(1);

    // menu blur
    fireEvent.blur(getByTestId('dropdown'));
    expect(queryAllByTestId('dropdown')).toHaveLength(0);

    // opener focus + enter keydown
    fireEvent.focus(getByTestId('text-button'));
    fireEvent.keyDown(getByTestId('text-button'), { key: 'Enter' });
    expect(queryAllByTestId('dropdown')).toHaveLength(1);

    // opener enter keydown again
    fireEvent.keyDown(getByTestId('text-button'), { key: 'Enter' });
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
  });

  it('should render loading state', () => {
    const { queryAllByTestId, getByTestId } = render(
      <Select loading>
        <DropdownItem value='1'>One</DropdownItem>
        <DropdownItem value='2'>Two</DropdownItem>
        <DropdownItem value='3'>Three</DropdownItem>
      </Select>,
    );

    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('spinner')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);

    fireEvent.mouseDown(getByTestId('field:block'));
    expect(queryAllByTestId('dropdown')).toHaveLength(1);
    expect(queryAllByTestId('spinner')).toHaveLength(1);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);
  });

  it('should handle menu keyboard control', () => {
    const spy = jest.fn();

    const { queryAllByTestId, getByTestId } = render(
      <Select onChange={spy}>
        <DropdownItem value='1'>One</DropdownItem>
        <DropdownItem value='2'>Two</DropdownItem>
        <DropdownItem value='3'>Three</DropdownItem>
      </Select>,
    );

    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);

    // show menu
    fireEvent.mouseDown(getByTestId('field:block'));
    expect(queryAllByTestId('dropdown')).toHaveLength(1);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(3);

    // make options active: 1, 2, 3, 1, 3, 2
    fireEvent.keyDown(getByTestId('dropdown'), { key: 'ArrowDown' });
    fireEvent.keyDown(getByTestId('dropdown'), { key: 'ArrowDown' });
    fireEvent.keyDown(getByTestId('dropdown'), { key: 'ArrowDown' });
    fireEvent.keyDown(getByTestId('dropdown'), { key: 'ArrowDown' });
    fireEvent.keyDown(getByTestId('dropdown'), { key: 'ArrowUp' });
    fireEvent.keyDown(getByTestId('dropdown'), { key: 'ArrowUp' });
    fireEvent.keyDown(getByTestId('dropdown'), { key: 'Enter' });
    expect(spy).toBeCalledTimes(1);
    expect(spy.mock.calls[0][0]).toEqual({ value: '2' });
  });

  it('should handle menu mouse control', () => {
    const spy = jest.fn();

    const { queryAllByTestId, getByTestId } = render(
      <Select onChange={spy}>
        <DropdownItem value='1'>One</DropdownItem>
        <DropdownItem value='2'>Two</DropdownItem>
        <DropdownItem value='3'>Three</DropdownItem>
      </Select>,
    );

    expect(spy).toBeCalledTimes(0);
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);

    fireEvent.mouseDown(getByTestId('field:block'));
    expect(spy).toBeCalledTimes(0);
    expect(queryAllByTestId('dropdown')).toHaveLength(1);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(3);

    fireEvent.click(queryAllByTestId('dropdown-item')[1]);
    expect(spy).toBeCalledTimes(1);
    expect(spy.mock.calls[0][0]).toEqual({ value: '2' });
    expect(queryAllByTestId('dropdown')).toHaveLength(0);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(0);
  });

  it('should handle options without defined value', () => {
    const spy = jest.fn();

    const { queryAllByTestId, getByTestId } = render(
      <Select onChange={spy}>
        <DropdownItem>Foo</DropdownItem>
        <DropdownItem>Bar</DropdownItem>
        <DropdownItem>Baz</DropdownItem>
      </Select>,
    );

    // open menu
    fireEvent.mouseDown(getByTestId('field:block'));
    expect(spy).toBeCalledTimes(0);
    expect(queryAllByTestId('dropdown')).toHaveLength(1);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(3);

    // click on third option
    fireEvent.click(queryAllByTestId('dropdown-item')[2]);
    expect(spy).toBeCalledTimes(1);
    expect(spy.mock.calls[0][0]).toEqual({ value: 'Baz' });
  });

  it('should handle blur on opener', () => {
    const { getByTestId } = render(
      <Select>
        <DropdownItem>Foo</DropdownItem>
        <DropdownItem>Bar</DropdownItem>
        <DropdownItem>Baz</DropdownItem>
      </Select>,
    );

    fireEvent.focus(getByTestId('field:block'));

    expect(() => {
      fireEvent.blur(getByTestId('field:block'));
    }).not.toThrow();
  });

  it('should handle menu enter keydown without active options', () => {
    const spy = jest.fn();

    const { getByTestId, queryAllByTestId } = render(
      <Select onChange={spy}>
        <DropdownItem>Foo</DropdownItem>
        <DropdownItem>Bar</DropdownItem>
        <DropdownItem>Baz</DropdownItem>
      </Select>,
    );

    // show menu
    fireEvent.mouseDown(getByTestId('field:block'));
    expect(queryAllByTestId('dropdown')).toHaveLength(1);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(3);

    // enter without active options
    fireEvent.keyDown(getByTestId('dropdown'), { key: 'Enter' });
    expect(spy).toBeCalledTimes(0);
  });

  it('should handle custom opener', () => {
    const spy = jest.fn();

    function CustomOpener({
      openerRef,
      onFocus,
      onBlur,
      onKeyDown,
      onMouseDown,
    }: SelectOpenerProps) {
      return (
        <div
          ref={openerRef as any}
          tabIndex={0}
          role='combobox'
          data-testid='custom-opener'
          {...{
            onFocus,
            onBlur,
            onKeyDown,
            onMouseDown,
          }}
        >
          My Opener
        </div>
      );
    }

    const { getByTestId, queryAllByTestId } = render(
      <Select opener={CustomOpener} onChange={spy}>
        <DropdownItem>Foo</DropdownItem>
        <DropdownItem>Bar</DropdownItem>
        <DropdownItem>Baz</DropdownItem>
      </Select>,
    );

    // show menu
    fireEvent.mouseDown(getByTestId('custom-opener'));
    expect(queryAllByTestId('dropdown')).toHaveLength(1);
    expect(queryAllByTestId('dropdown-item')).toHaveLength(3);

    // click on third option
    fireEvent.click(queryAllByTestId('dropdown-item')[2]);
    expect(spy).toBeCalledTimes(1);
    expect(spy.mock.calls[0][0]).toEqual({ value: 'Baz' });
  });
});

describe('SelectContext', () => {
  it('should provide defaults', () => {
    function TestComponent() {
      const { setDropdownProps } = useContext(SelectContext);

      return (
        <div data-testid='test-component' onClick={() => setDropdownProps({ id: 'some-id' })}>
          Hello
        </div>
      );
    }

    const { getByTestId } = render(<TestComponent />);

    expect(() => {
      fireEvent.click(getByTestId('test-component'));
    }).not.toThrow();
  });
});
