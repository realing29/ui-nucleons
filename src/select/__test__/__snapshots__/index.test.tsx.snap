// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render loading state 1`] = `
<Select
  label="Формат каталога"
  loading={true}
  options={
    Array [
      "Прайс-лист XML",
      "Прайс-лист YML",
      "Карточки товаров JPG",
      "Фото товаров JPG + прайс-лист XLS",
      "Печатный каталог PDF",
    ]
  }
  value="Карточки товаров JPG"
>
  <div
    className="root"
  >
    <ForwardRef(TextField)
      className="field"
      data-testid="select:field"
      endAdornment={
        <UpSVG
          fill="#9e9e9e"
        />
      }
      focused={true}
      label="Формат каталога"
      multiline={false}
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      readOnly={true}
      value="Карточки товаров JPG"
      variant="desktop"
    >
      <div
        className="root field"
        data-testid="select:field"
      >
        <div
          className="reset input-block focused variant-desktop size-l with-label rounds-sm__all"
          data-testid="text-field:block"
          onClick={[Function]}
        >
          <div
            className="main-column"
          >
            <label
              className="label focused variant-desktop filled"
            >
              Формат каталога
            </label>
            <div
              className="input-wrapper"
            >
              <ForwardRef(BaseInput)
                className="base-input focused variant-desktop size-l"
                data-testid="text-field:field"
                multiline={false}
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={true}
                value="Карточки товаров JPG"
              >
                <div
                  className="reset root base-input focused variant-desktop size-l"
                >
                  <input
                    className="field reset"
                    data-testid="text-field:field"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                    onKeyDown={[Function]}
                    readOnly={true}
                    value="Карточки товаров JPG"
                  />
                </div>
              </ForwardRef(BaseInput)>
            </div>
          </div>
          <Box
            alignItems="center"
            display="flex"
            marginLeft={4}
          >
            <div
              className="display-flex align-items-center M-l__4"
            >
              <UpSVG
                fill="#9e9e9e"
              >
                <svg
                  fill="#9e9e9e"
                  height={16}
                  viewBox="0 0 16 16"
                  width={16}
                >
                  <path
                    clipRule="evenodd"
                    d="M12.753 9.659a1 1 0 01-1.412.094L8 6.829 4.659 9.753a1 1 0 01-1.317-1.506l4-3.5a1 1 0 011.317 0l4 3.5a1 1 0 01.094 1.412z"
                    fillRule="evenodd"
                  />
                </svg>
              </UpSVG>
            </div>
          </Box>
        </div>
      </div>
    </ForwardRef(TextField)>
    <ForwardRef(Dropdown)
      className="menu"
      data-testid="select:menu"
      onBlur={[Function]}
      onKeyDown={[Function]}
      role="menu"
      style={
        Object {
          "top": "56px",
        }
      }
      tabIndex={-1}
    >
      <div
        className="root menu shadow-z4 rounds-md__all"
        data-testid="select:menu"
        onBlur={[Function]}
        onKeyDown={[Function]}
        role="menu"
        style={
          Object {
            "top": "56px",
          }
        }
        tabIndex={-1}
      >
        <CustomScrollbar
          className="inner"
          overflow={
            Object {
              "x": "h",
              "y": "s",
            }
          }
        >
          <OverlayScrollbarsComponent
            className="custom-scrollbar inner"
            options={
              Object {
                "callbacks": undefined,
                "overflowBehavior": Object {
                  "x": "h",
                  "y": "s",
                },
              }
            }
          >
            <div
              className="os-host"
            >
              <div
                className="os-resize-observer-host"
              />
              <div
                className="os-padding"
              >
                <div
                  className="os-viewport"
                >
                  <div
                    className="os-content"
                  >
                    <DropdownLoading
                      data-testid="select:loading-area"
                    >
                      <div
                        className="root"
                        data-testid="select:loading-area"
                      >
                        <Spinner
                          size="small"
                        >
                          <div>
                            <svg
                              className="spinner size-small"
                              height={20}
                              stroke="#1f84db"
                              viewBox="0 0 20 20"
                              width={20}
                            >
                              <circle
                                className="circle"
                                cx={10}
                                cy={10}
                                r={9}
                              />
                            </svg>
                          </div>
                        </Spinner>
                      </div>
                    </DropdownLoading>
                  </div>
                </div>
              </div>
              <div
                className="os-scrollbar os-scrollbar-horizontal "
              >
                <div
                  className="os-scrollbar-track"
                >
                  <div
                    className="os-scrollbar-handle"
                  />
                </div>
              </div>
              <div
                className="os-scrollbar os-scrollbar-vertical"
              >
                <div
                  className="os-scrollbar-track"
                >
                  <div
                    className="os-scrollbar-handle"
                  />
                </div>
              </div>
              <div
                className="os-scrollbar-corner"
              />
            </div>
          </OverlayScrollbarsComponent>
        </CustomScrollbar>
      </div>
    </ForwardRef(Dropdown)>
  </div>
</Select>
`;

exports[`Select should renders correctly 1`] = `
<Select
  label="Формат каталога"
  onClick={[MockFunction]}
  onMenuToggle={
    [MockFunction] {
      "calls": Array [
        Array [
          false,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onSelect={[MockFunction]}
  optionSize="m"
  options={
    Array [
      "Прайс-лист XML",
      "Прайс-лист YML",
      "Карточки товаров JPG",
      "Фото товаров JPG + прайс-лист XLS",
      "Печатный каталог PDF",
    ]
  }
  renderOption={[Function]}
  value="Карточки товаров JPG"
>
  <div
    className="root"
  >
    <ForwardRef(TextField)
      className="field"
      data-testid="select:field"
      endAdornment={
        <DownSVG
          fill="#9e9e9e"
        />
      }
      focused={false}
      label="Формат каталога"
      multiline={false}
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      readOnly={true}
      value="Карточки товаров JPG"
      variant="desktop"
    >
      <div
        className="root field"
        data-testid="select:field"
      >
        <div
          className="reset input-block variant-desktop size-l with-label rounds-sm__all"
          data-testid="text-field:block"
          onClick={[Function]}
        >
          <div
            className="main-column"
          >
            <label
              className="label variant-desktop filled"
            >
              Формат каталога
            </label>
            <div
              className="input-wrapper"
            >
              <ForwardRef(BaseInput)
                className="base-input variant-desktop size-l"
                data-testid="text-field:field"
                multiline={false}
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={true}
                value="Карточки товаров JPG"
              >
                <div
                  className="reset root base-input variant-desktop size-l"
                >
                  <input
                    className="field reset"
                    data-testid="text-field:field"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                    onKeyDown={[Function]}
                    readOnly={true}
                    value="Карточки товаров JPG"
                  />
                </div>
              </ForwardRef(BaseInput)>
            </div>
          </div>
          <Box
            alignItems="center"
            display="flex"
            marginLeft={4}
          >
            <div
              className="display-flex align-items-center M-l__4"
            >
              <DownSVG
                fill="#9e9e9e"
              >
                <svg
                  fill="#9e9e9e"
                  height={16}
                  viewBox="0 0 16 16"
                  width={16}
                >
                  <path
                    clipRule="evenodd"
                    d="M3.247 6.342a1 1 0 011.412-.095L8 9.171l3.341-2.924a1 1 0 011.318 1.506l-4 3.5a1 1 0 01-1.317 0l-4-3.5a1 1 0 01-.095-1.411z"
                    fillRule="evenodd"
                  />
                </svg>
              </DownSVG>
            </div>
          </Box>
        </div>
      </div>
    </ForwardRef(TextField)>
  </div>
</Select>
`;

exports[`Select should renders correctly 2`] = `
<Select
  label="Формат каталога"
  onClick={[MockFunction]}
  onMenuToggle={
    [MockFunction] {
      "calls": Array [
        Array [
          false,
        ],
        Array [
          true,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onSelect={[MockFunction]}
  optionSize="m"
  options={
    Array [
      "Прайс-лист XML",
      "Прайс-лист YML",
      "Карточки товаров JPG",
      "Фото товаров JPG + прайс-лист XLS",
      "Печатный каталог PDF",
    ]
  }
  renderOption={[Function]}
  value="Карточки товаров JPG"
>
  <div
    className="root"
  >
    <ForwardRef(TextField)
      className="field"
      data-testid="select:field"
      endAdornment={
        <UpSVG
          fill="#9e9e9e"
        />
      }
      focused={true}
      label="Формат каталога"
      multiline={false}
      onBlur={[Function]}
      onClick={[Function]}
      onKeyDown={[Function]}
      readOnly={true}
      value="Карточки товаров JPG"
      variant="desktop"
    >
      <div
        className="root field"
        data-testid="select:field"
      >
        <div
          className="reset input-block focused variant-desktop size-l with-label rounds-sm__all"
          data-testid="text-field:block"
          onClick={[Function]}
        >
          <div
            className="main-column"
          >
            <label
              className="label focused variant-desktop filled"
            >
              Формат каталога
            </label>
            <div
              className="input-wrapper"
            >
              <ForwardRef(BaseInput)
                className="base-input focused variant-desktop size-l"
                data-testid="text-field:field"
                multiline={false}
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={true}
                value="Карточки товаров JPG"
              >
                <div
                  className="reset root base-input focused variant-desktop size-l"
                >
                  <input
                    className="field reset"
                    data-testid="text-field:field"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                    onKeyDown={[Function]}
                    readOnly={true}
                    value="Карточки товаров JPG"
                  />
                </div>
              </ForwardRef(BaseInput)>
            </div>
          </div>
          <Box
            alignItems="center"
            display="flex"
            marginLeft={4}
          >
            <div
              className="display-flex align-items-center M-l__4"
            >
              <UpSVG
                fill="#9e9e9e"
              >
                <svg
                  fill="#9e9e9e"
                  height={16}
                  viewBox="0 0 16 16"
                  width={16}
                >
                  <path
                    clipRule="evenodd"
                    d="M12.753 9.659a1 1 0 01-1.412.094L8 6.829 4.659 9.753a1 1 0 01-1.317-1.506l4-3.5a1 1 0 011.317 0l4 3.5a1 1 0 01.094 1.412z"
                    fillRule="evenodd"
                  />
                </svg>
              </UpSVG>
            </div>
          </Box>
        </div>
      </div>
    </ForwardRef(TextField)>
    <ForwardRef(Dropdown)
      className="menu"
      data-testid="select:menu"
      onBlur={[Function]}
      onKeyDown={[Function]}
      role="menu"
      style={
        Object {
          "top": "56px",
        }
      }
      tabIndex={-1}
    >
      <div
        className="root menu shadow-z4 rounds-md__all"
        data-testid="select:menu"
        onBlur={[Function]}
        onKeyDown={[Function]}
        role="menu"
        style={
          Object {
            "top": "56px",
          }
        }
        tabIndex={-1}
      >
        <CustomScrollbar
          className="inner"
          overflow={
            Object {
              "x": "h",
              "y": "s",
            }
          }
        >
          <OverlayScrollbarsComponent
            className="custom-scrollbar inner"
            options={
              Object {
                "callbacks": undefined,
                "overflowBehavior": Object {
                  "x": "h",
                  "y": "s",
                },
              }
            }
          >
            <div
              className="os-host"
            >
              <div
                className="os-resize-observer-host"
              />
              <div
                className="os-padding"
              >
                <div
                  className="os-viewport"
                >
                  <div
                    className="os-content"
                  >
                    <DropdownItem
                      key="0"
                      onClick={[Function]}
                      role="menuitem"
                      size="m"
                    >
                      <div
                        className="root size-m"
                        data-testid="dropdown-item"
                        onClick={[Function]}
                        role="menuitem"
                      >
                        ПРАЙС-ЛИСТ XML
                      </div>
                    </DropdownItem>
                    <DropdownItem
                      key="1"
                      onClick={[Function]}
                      role="menuitem"
                      size="m"
                    >
                      <div
                        className="root size-m"
                        data-testid="dropdown-item"
                        onClick={[Function]}
                        role="menuitem"
                      >
                        ПРАЙС-ЛИСТ YML
                      </div>
                    </DropdownItem>
                    <DropdownItem
                      key="2"
                      onClick={[Function]}
                      role="menuitem"
                      size="m"
                    >
                      <div
                        className="root size-m"
                        data-testid="dropdown-item"
                        onClick={[Function]}
                        role="menuitem"
                      >
                        КАРТОЧКИ ТОВАРОВ JPG
                      </div>
                    </DropdownItem>
                    <DropdownItem
                      key="3"
                      onClick={[Function]}
                      role="menuitem"
                      size="m"
                    >
                      <div
                        className="root size-m"
                        data-testid="dropdown-item"
                        onClick={[Function]}
                        role="menuitem"
                      >
                        ФОТО ТОВАРОВ JPG + ПРАЙС-ЛИСТ XLS
                      </div>
                    </DropdownItem>
                    <DropdownItem
                      key="4"
                      onClick={[Function]}
                      role="menuitem"
                      size="m"
                    >
                      <div
                        className="root size-m"
                        data-testid="dropdown-item"
                        onClick={[Function]}
                        role="menuitem"
                      >
                        ПЕЧАТНЫЙ КАТАЛОГ PDF
                      </div>
                    </DropdownItem>
                  </div>
                </div>
              </div>
              <div
                className="os-scrollbar os-scrollbar-horizontal "
              >
                <div
                  className="os-scrollbar-track"
                >
                  <div
                    className="os-scrollbar-handle"
                  />
                </div>
              </div>
              <div
                className="os-scrollbar os-scrollbar-vertical"
              >
                <div
                  className="os-scrollbar-track"
                >
                  <div
                    className="os-scrollbar-handle"
                  />
                </div>
              </div>
              <div
                className="os-scrollbar-corner"
              />
            </div>
          </OverlayScrollbarsComponent>
        </CustomScrollbar>
      </div>
    </ForwardRef(Dropdown)>
  </div>
</Select>
`;
